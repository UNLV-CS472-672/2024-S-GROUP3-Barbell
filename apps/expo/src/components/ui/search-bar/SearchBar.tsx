/**
 * HOW TO USE:
 * --------------------------------------------------------------------------------
 * Inside of your PARENT component, you need to have the following:
 * - list: any[]
 *    - This is your raw data that you get from a query
 * - const[filteredList, setFilteredList] = useState<any[]>(List)
 *    - filteredList is what is going to be displayed
 *    - setFilteredList is how we set the values that are going to be outputted
 * - filterBy: string
 *    - this is the field that you want to filter your data by
 * - placeholder: string
 *    - placeholder text within the search bar
 * --------------------------------------------------------------------------------
 * To output the data, you want to use a map and map the items within filteredList
 * to your list item component.
 *
 * For example:
 *  <SearchBar list={list} placeholder="Search..." setFilteredList={setFilteredList} filterBy="name" />
 *  {filteredList.map((item, index) => (
 *    <View key={index}>
 *      <Text>{item.name}</Text>
 *    </View>
 *  ))}
 */

import React, { useEffect, useState } from 'react'
import { View } from 'react-native'
import { TextInput } from 'react-native-gesture-handler'

import { Fontisto } from '@expo/vector-icons'

interface SearchBarProps {
  list: any[] | undefined
  setFilteredList: React.Dispatch<React.SetStateAction<any[] | undefined>>
  filterBy: string
  placeholder: string
}

export default function SearchBar({
  list,
  setFilteredList,
  filterBy,
  placeholder,
}: SearchBarProps) {
  const [searchTerm, setSearchTerm] = useState('')

  useEffect(() => {
    /* istanbul ignore next */
    if (list) {
      setFilteredList(
        list.filter((item) => item[filterBy].toLowerCase().includes(searchTerm.toLowerCase())),
      )
    }
  }, [list, filterBy, searchTerm])

  return (
    <View
      className='mx-3 my-2 flex-row items-center rounded-xl bg-[#272727] px-3'
      testID='searchBar'
    >
      <Fontisto name='search' size={24} color='#717171' />
      <TextInput
        className='placeholder:color-[#717171] w-full px-3 py-3 text-[20px] text-white'
        placeholder={placeholder}
        onChangeText={setSearchTerm}
      />
    </View>
  )
}
