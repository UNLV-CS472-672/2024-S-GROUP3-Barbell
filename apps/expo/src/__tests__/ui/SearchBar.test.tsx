import React, { useState } from 'react'

import { fireEvent, render, screen } from '@testing-library/react-native'

import SearchBar from '~/components/ui/search-bar/SearchBar'

test('SearchBar', async () => {
  const TestComponent = () => {
    const item1 = {
      id: 1,
      name: 'ab',
    }
    const item2 = {
      id: 2,
      name: 'abc',
    }
    const item3 = {
      id: 3,
      name: 'bac',
    }

    const listToFilter = [item1, item2, item3]
    const [filteredList, setFilteredList] = useState<any[] | undefined>(listToFilter)
    return <SearchBar list={listToFilter} placeholder="Search..." setFilteredList={setFilteredList} filterBy="name" />
  }

  render(<TestComponent />)

  // Find the search input
  const searchInput = screen.getByPlaceholderText('Search...')

  // Simulate typing in the search input
  fireEvent.changeText(searchInput, 'a') // Type 'a'
})
