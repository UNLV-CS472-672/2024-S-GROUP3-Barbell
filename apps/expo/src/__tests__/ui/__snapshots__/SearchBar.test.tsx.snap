// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Component renders correctly and matches snapshot 1`] = `
<View
  className="mx-3 my-2 flex-row items-center rounded-[5px] bg-[#272727] px-3"
  testID="searchBar"
>
  <Text />
  <TextInput
    className="placeholder:color-[#717171] px-3 py-[6px] text-[20px] text-white"
    collapsable={false}
    handlerTag={1}
    handlerType="NativeViewGestureHandler"
    onChangeText={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    placeholder="Search..."
  />
</View>
`;
